libs/utils/src/shared_retryable.rs

```diff
@@ -0,0 +1,657 @@
+/// "inner error" which will be used to build an outer error. The outer error will also have to be
+/// convertable from [`RetriedTaskPanicked`] to absorb that case as well.
+///
+/// ## Example
+///
+/// A shared service value completes the infallible work once, even if called concurrently by
+/// multiple cancellable tasks.
+///
+/// Example moved as a test `service_example`.
+#[derive(Clone)]
+pub struct SharedRetryable<V> {
+    inner: Arc<tokio::sync::Mutex<Option<MaybeDone<V>>>>,
+}
+
+impl<V> Default for SharedRetryable<V> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(tokio::sync::Mutex::new(None)),
+        }
+    }
+}
+
+/// Determine if an error is transient or permanent.
+pub trait Retryable {
+    fn is_permanent(&self) -> bool {
+        true
+    }
+}
+
+/// Retried task panicked, was cancelled, or never spawned (see [`SharedRetryable::try_restart`]).
+#[derive(Debug, PartialEq, Eq)]
+pub struct RetriedTaskPanicked;
+
+impl<T, E1> SharedRetryable<Result<T, E1>>
+where
+    T: Clone + std::fmt::Debug + Send + 'static,
+    E1: Clone + Retryable + std::fmt::Debug + Send + 'static,
+{
+    /// Restart a previously failed operation unless it already completed with a terminal result.
+    ///
+    /// Many futures can call this function and and get the terminal result from an earlier attempt
+    /// or start a new attempt, or join an existing one.
+    ///
+    /// Compared to `Self::try_restart`, this method also spawns the future to run, which would
+    /// otherwise have to be done manually.
+    #[cfg(test)]
+    pub async fn try_restart_spawn<F, Fut, E2>(&self, retry_with: F) -> Result<T, E2>
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = Result<T, E1>> + Send + 'static,
+        E2: From<E1> + From<RetriedTaskPanicked> + Send + 'static,
+    {
+        let (recv, maybe_fut) = self.try_restart(retry_with).await;
+
+        if let Some(fut) = maybe_fut {
+            // top level function, we must spawn, pageserver cannot use this
+            tokio::spawn(fut);
+        }
+
+        recv.await
+    }
+
+    /// Restart a previously failed operation unless it already completed with a terminal result.
+    ///
+    /// Many futures can call this function and get the terminal result from an earlier attempt or
+    /// start a new attempt, or join an existing one.
+    ///
+    /// If a task calling this method is cancelled, at worst, a new attempt which is immediatedly
+    /// deemed as having panicked will happen, but without a panic ever happening.
+    ///
+    /// Returns one future for waiting for the result and possibly another which needs to be
+    /// spawned when `Some`. Spawning has to happen before waiting is started, otherwise the first
+    /// future will never make progress.
+    ///
+    /// This complication exists because on pageserver we cannot use `tokio::spawn` directly
+    /// at this time.
```

Comment:
Spawner requires boxing at the caller, so I'd prefer not doing that, it is what SpawnExt does: https://docs.rs/futures-util/0.3.28/src/futures_util/task/spawn.rs.html#49-51

Discussed this also with @funbringer, we could have a trait+ZST for spawning on tokio or on a closure.

Nice thing that followed after not spawning deep in the callstack is that the easiest way now to accidentially "leak" the `strong: Arc<Receiver<_>>` requires `std::mem::forget` which can hardly be accidential. It is probably not obvious looking at the latest iteration but that's why I don't really like to revert back.

There are tests to show that dropping the future needing to be spawned will produce a failure. I guess it could be a fast looping situation, but that should really come up during local testing I think.

I could add some `#[must_use]` to `try_restart`. Would that help? I didn't check if they do get implicit `#[must_use]` from `std::future::Future`.

Reply:
> Spawner requires boxing at the caller, so I'd prefer not doing that, it is what SpawnExt does:

Yep, thats why I wrote this: `Probably our own, not this particular one from futures`. 

> we could have a trait+ZST for spawning on tokio or on a closure.

Was thinking about something of that nature. 

My point is that internal spawning should simplify signatures a lot, i e no need to return multiple `Future<Output=...>`. 

In general would be interesting to see how this decision affects complexity of the whole primitive, but personally I dont have capacity to run such an experiment, so I'm on board with current approach.

> I could add some #[must_use]

Having `#[must_use]` would be a plus.